Kernel/Process.cpp

```diff
@@ -63,6 +63,67 @@ SpinlockProtected<Process::List>& Process::all_instances()
     return *s_all_instances;
 }
 
+ErrorOr<void> Process::for_each_with_jail_respect(Function<ErrorOr<void>(Process&)> callback)
+{
+    ErrorOr<void> result {};
+    jail().with([&](auto my_jail) {
+        Process::all_instances().with([&](auto const& list) {
+            for (auto it = list.begin(); it != list.end();) {
+                auto& process = *it;
+                ++it;
+                process.jail().with([&](auto& their_jail) {
+                    if (!my_jail || their_jail.ptr() == my_jail.ptr())
+                        result = callback(process);
+                });
+                if (result.is_error())
+                    break;
+            }
+        });
+    });
+    return result;
+}
+
+ErrorOr<void> Process::for_each_child_with_jail_respect(Function<ErrorOr<void>(Process&)> callback)
+{
+    ProcessID my_pid = pid();
+    ErrorOr<void> result {};
+    jail().with([&](auto my_jail) {
+        Process::all_instances().with([&](auto const& list) {
+            for (auto it = list.begin(); it != list.end();) {
+                auto& process = *it;
+                ++it;
+                process.jail().with([&](auto& their_jail) {
+                    if ((!my_jail || their_jail.ptr() == my_jail.ptr()) && (process.ppid() == my_pid || process.has_tracee_thread(pid())))
```

Comment:
here and elsewhere, we don't need to take another process's jail lock to know that we don't have a jail. I feel like even though it may be duplicated, it will be useful to split this out into two branches where we don't bother taking the 3rd level of lock when we don't have our own jail.